src/test/java/org/example/springbootintro/category/CategoryControllerTest.java

```diff
@@ -0,0 +1,180 @@
+        requestDto.setDescription("TestDescription");
+
+        CategoryDto expectedCategoryDto = new CategoryDto();
+        expectedCategoryDto.setId(1L);
+        expectedCategoryDto.setName("TestCategory");
+        expectedCategoryDto.setDescription("TestDescription");
+
+        given(categoryService.save(requestDto)).willReturn(expectedCategoryDto);
+
+        // When
+        mockMvc.perform(post("/categories")
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(UtilsTest.asJsonString(requestDto)))
+                // Then
+                .andExpect(status().isCreated())
+                .andExpect(jsonPath("$.id").value(1L))
+                .andExpect(jsonPath("$.name").value("TestCategory"))
+                .andExpect(jsonPath("$.description").value("TestDescription"));
+
+        // Verify
+        verify(categoryService).save(requestDto);
+        verifyNoMoreInteractions(categoryService);
+    }
+
+    @Test
+    @DisplayName("Test find category by id")
+    public void findCategoryByIdTest() throws Exception {
+        // Given
+        CategoryDto categoryDto = new CategoryDto(1L, "TestCategory", "TestDescription");
+        given(categoryService.findById(1L)).willReturn(categoryDto);
+
+        // When
+        mockMvc.perform(get("/categories/{id}", 1L)
+                        .contentType(MediaType.APPLICATION_JSON))
+                // Then
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.id").value(1L))
+                .andExpect(jsonPath("$.name").value("TestCategory"))
+                .andExpect(jsonPath("$.description").value("TestDescription"));
+
+        // Verify
+        verify(categoryService).findById(1L);
+        verifyNoMoreInteractions(categoryService);
+    }
+
+    @Test
+    @WithMockUser(roles = "ADMIN")
+    @DisplayName("Test update category by id")
+    public void updateCategoryByIdTest() throws Exception {
+        // Given
+        CreateCategoryRequestDto requestDto = new CreateCategoryRequestDto();
+        requestDto.setName("TestCategory");
+        requestDto.setDescription("TestDescription");
+
+        CategoryDto expectedCategoryDto = new CategoryDto();
+        expectedCategoryDto.setId(1L);
+        expectedCategoryDto.setName("TestCategory");
+        expectedCategoryDto.setDescription("TestDescription");
+
+        given(categoryService.updateById(1L, requestDto)).willReturn(expectedCategoryDto);
+
+        // When
+        mockMvc.perform(put("/categories/{id}", 1L)
+                        .contentType(MediaType.APPLICATION_JSON)
+                        .content(UtilsTest.asJsonString(requestDto)))
+                // Then
+                .andExpect(status().isOk())
+                .andExpect(jsonPath("$.id").value(1L))
+                .andExpect(jsonPath("$.name").value("TestCategory"))
+                .andExpect(jsonPath("$.description").value("TestDescription"));
```

Comment:
this when block also contains an assertion of the returned data, so it is not pure `when` section, would be nice to have a more clear separation of actions, check other similar places as well